### PR TITLE
How to Refactor and Not Break Things

### PR DESCRIPTION
**Pull request by antonpirker**
A blog post about how we did a big refactoring without breaking things. 

Preview: https://sentry-engineering-git-antonpirker-refactoring.sentry.dev/blog/how-to-refactor-and-not-break-things

### CODE REVIEW

File: data/blog/how-to-refactor-and-not-break-things.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+images: ['/images/how-to-refactor-and-not-break-things/hero.jpg']
+layout: PostLayout
+authors: ['antonpirker']
+---
+
+In our Python SDK, we completed a huge refactoring, and I want to write down how we pulled this off without breaking (almost) anything and how we managed to stay mostly backward compatible.
+
+## The Initial Situation
+
+When you add the Sentry SDK to your project it hooks into the execution of your application and collects data during execution of your application. Depending on the frameworks or libraries you use the data is collected in different parts of your project and is sent to Sentry in other parts.
```
Comment (bruno-garcia) [inline data/blog/how-to-refactor-and-not-break-things.md:16]: > it hooks into the execution of your application and collects data during execution of your application.

This is hard to parse. I can't quite give a suggestion though, what are you trying to say exactly?

Something like?
```suggestion
When you add Sentry to your application, it'll instrument the app and runtime to collect useful debugging information at runtime. Depending on the frameworks or libraries you use, the data is collected at different stages of process execution, and is sent to Sentry at a later stage, asynchronously. 
```

File: data/blog/how-to-refactor-and-not-break-things.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+layout: PostLayout
+authors: ['antonpirker']
+---
+
+In our Python SDK, we completed a huge refactoring, and I want to write down how we pulled this off without breaking (almost) anything and how we managed to stay mostly backward compatible.
+
+## The Initial Situation
+
+When you add the Sentry SDK to your project it hooks into the execution of your application and collects data during execution of your application. Depending on the frameworks or libraries you use the data is collected in different parts of your project and is sent to Sentry in other parts. 
+If your project consists of multiple services (like the frontend, the backend, some microservices, or worker processes consuming items from a queue) the Sentry SDK also propagates tracing information between those services. This makes it possible to link the data from all your services into one trace.
```
Comment (bruno-garcia) [inline data/blog/how-to-refactor-and-not-break-things.md:17]: Should we call it application? or system? instead of project? Not sure if it's just me but, "Sentry Project" is a Sentry thing, but when I think of my projects, I don't necessarily think of an app, or a larger system, or product. 

```suggestion
If your application consists of multiple services (like frontend, backend, some microservices, or worker processes consuming items from a queue) the Sentry SDK also propagates tracing information between those services. This makes it possible to link the data from all your services into one trace.
```

File: data/blog/how-to-refactor-and-not-break-things.md [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+To handle all this data the SDK uses a thing called the Hub. The Hub holds a stack of so-called Scopes. All this was specified in the [Unified API](https://develop.sentry.dev/sdk/unified-api/) a long time ago. Data is sent to Sentry as events. Before sending events to Sentry the data from the Scope(s) is applied to those events. The SDK needs to make sure that only data from the right Scopes is applied to not leak data for example from one thread into the events captured by another thread.
```
Comment (bruno-garcia) [inline data/blog/how-to-refactor-and-not-break-things.md:19]:
```suggestion
To handle all this data the SDK uses a thing called the Hub. The Hub holds a stack of so-called Scopes. All this was specified in the [Unified API](https://develop.sentry.dev/sdk/unified-api/) a long time ago. Data is sent to Sentry as events. Before sending events to Sentry the data from the Scope(s) is applied to those events. The SDK needs to make sure that only data from the right Scopes is applied to not leak data for example from one thread into the events captured by another thread. Think of a each web request having its own `url` tag, or `user.id`.
```

File: data/blog/how-to-refactor-and-not-break-things.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+When doing a big refactoring like this the WHY is very important. We do not want to refactor just for refactoring's sake. 
+In recent years [OpenTelementry](https://opentelemetry.io/) has become more popular and we wanted to make sure Sentry is compatible with OpenTelemetry. 
+We discovered that with our current Hub implementation that was not the case, so we decided to remove the Hub and switch to have only Scopes and make them behave like Contexts in OpenTelementry. This makes Sentry 100% compatible with OpenTelemetry and paves the way for the future.
+
+## This Is a Huge Undertaking, We need a Plan!
+
+Some goals that we wanted to achieve with the refactoring:
+
+- Be as backward compatible as possible. If possible, our users should not need to change their custom Sentry code.
+- Because we changed how we handle data, we wanted to make sure that we do not use more resources (CPU and memory) compared to before the refactoring.
```
Comment (bruno-garcia) [inline data/blog/how-to-refactor-and-not-break-things.md:36]: this could be simplified, e.g: "Avoid introducing additional overhead"

File: data/blog/how-to-refactor-and-not-break-things.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+After we moved all the functionality from the Hub into the Scope our extensive test suite was still all green, giving us confidence that we did not change any behavior. Because neither the top-level Sentry API nor the Hub-based API was changed this could be released in a minor version update.
+
+This concluded the preparation of our canvas. Time to make the change.
+
+## Phase II: New Scopes
+
+Phase I left us with a hollow shell of a Hub and all functionality in the Scope. We now refactored the Scope. This was the biggest part. We changed how the Scope was stored in memory (it's not on the Hub anymore but saved as a Python Context Variable). We also introduced three different flavors of the Scope for better encapsulation of data. If you want to dig into details read our [develop docs on the topic](https://develop.sentry.dev/sdk/hub_and_scope_refactoring/).
+
+After we updated the Scope the old Hub-based API could still be used, but under the hood, the new Scopes-based API was called. We did not release this phase because we wanted to first do Phase III.
```
Comment (bruno-garcia) [inline data/blog/how-to-refactor-and-not-break-things.md:59]: > We did not release this phase because we wanted to first do Phase III.

This doesn't really explain why not. It wouldn't add value? Would it at least help make sure there were no bugs there? (beyond what the test suite can do of course)

File: data/blog/how-to-refactor-and-not-break-things.md [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+Again, our test suite gave us confidence that the SDK still behaved the same as before.
+
+## Phase III: Use New Scopes Everywhere
+
+In the Python SDK, we have over 40 integrations for various web frameworks, databases, and other libraries. All of them still used the Hub-based API.
+
+We updated each integration from the old Hub-based API to the new Scope-based API in a separate PR. This was pretty straightforward and done in a couple of days.
+
+Doing this gave us insights into the look and feel of the new API. Because we now used the API in the same way as our power users would use it in the future. We found some things that we did not like and made some minor changes to the new API to make it more convenient to use. Having one PR for each integration created a lot of small PRs that were easy to review.
```
Comment (bruno-garcia) [inline data/blog/how-to-refactor-and-not-break-things.md:69]: > We found some things that we did not like and made some minor changes to the new API to make it more convenient to use

Would love to see some code block here. What was it before and why didn't u like it?
What was it after?

File: data/blog/how-to-refactor-and-not-break-things.md [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+After we had all PRs merged, we created a release candidate and released it on PyPI so everyone could give it a try and we could gather feedback.
+
+We updated our internal usage dashboards so we could see the adoption of the new release candidate.
+
+## Phase IV: Load Testing and Dogfooding
+
+We started to do load testing. We created a [sample application and ran load tests](https://github.com/getsentry/demo-flask-load-test) locally to check if the CPU and memory usage would change between the current SDK and the new 2.0 version. It did not. After the load tests, we were confident that we could use the release candidate on Sentry.io.
+
+Sentry.io is a very big Python project where we use all the advanced features of the SDK. It is a perfect candidate for dogfooding because it will uncover problems very soon.
```
Comment (bruno-garcia) [inline data/blog/how-to-refactor-and-not-break-things.md:79]:
```suggestion
[Sentry.io is a very big Python code base](https://github.com/getsentry/sentry) where we use all the advanced features of the SDK. It is a perfect candidate for dogfooding because it will uncover problems very soon.
```

File: data/blog/how-to-refactor-and-not-break-things.md [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+We first installed the new SDK on our Canary servers. We let it run for an hour and closely monitored CPU and Memory. Everything looked good. Like with the local load tests, nothing spiked.
+
+Time for some proper dogfooding.
+
+We installed SDK 2.0 on **all** our servers. After running the new version for a week on Sentry.io we discovered that there was a problem in the Celery integration where the baggage header (used for propagating trace information) was handled incorrectly. After fixing this we created a new release candidate and continued the dogfooding.
```
Comment (bruno-garcia) [inline data/blog/how-to-refactor-and-not-break-things.md:85]: It's cool for those curious to link up to what the change involved:

```suggestion
We installed SDK 2.0 on **all** our servers. After running the new version for a week on Sentry.io we discovered that there was a problem in the Celery integration where the baggage header (used for propagating trace information) was handled incorrectly. [After fixing this](https://github.com/getsentry/sentry-python/pull/2993), we created a new release candidate and continued the dogfooding.
```

File: data/blog/how-to-refactor-and-not-break-things.md [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+All in all, we created 6 release candidates where we each time fixed some smaller problems or made improvements.
+
+## Finally: The Release
+
+When we were confident that all problems were fixed, all data was correct, and CPU and memory usage was the same as before we [released version 2.0](https://github.com/getsentry/sentry-python/releases/tag/2.0.0).
```
Comment (bruno-garcia) [inline data/blog/how-to-refactor-and-not-break-things.md:91]: > When we were confident that all problems were fixed

How long has it been out? 😄 
How about:

```suggestion
We fixed all issues we uncovered during dogfooding, giving us enough confidence to move forward. All data collected looked good, and CPU and memory usage was the same as before we [released version 2.0](https://github.com/getsentry/sentry-python/releases/tag/2.0.0).
```

File: data/blog/how-to-refactor-and-not-break-things.md [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+On the first day, there was a [bug report](https://github.com/getsentry/sentry-python/issues/3021) from a user when using the SDK with the Starlette framework and Uvicorn as the server. We fixed the problem and released [2.0.1](https://github.com/getsentry/sentry-python/releases/tag/2.0.1) right away.
```
Comment (bruno-garcia) [inline data/blog/how-to-refactor-and-not-break-things.md:93]: Oh OK so yeah it's better to reword the paragraph above indeed, since a patch was needed immediately.

File: data/blog/how-to-refactor-and-not-break-things.md [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+
+Since then the bug tracker has been quiet, and no other regressions were reported.
+
+Our internal usage dashboards showed that our users (the most prominent one being Spotify) were adopting the new major version and sent tens of millions of events to Sentry.
```
Comment (bruno-garcia) [inline data/blog/how-to-refactor-and-not-break-things.md:97]: I don't see them here: https://sentry.io/customers/ can we talk about them? please confirm that

```suggestion
Our internal usage dashboards showed that our users (like a very popular music streaming app) were adopting the new major version and sent tens of millions of events to Sentry.
```

**Review by bruno-garcia — COMMENTED** [review]
Thanks Anton, left some comments

File: data/blog/how-to-refactor-and-not-break-things.md [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+
+All in all, we created 6 release candidates where we each time fixed some smaller problems or made improvements.
+
+## Finally: The Release
+
+When we were confident that all problems were fixed, all data was correct, and CPU and memory usage was the same as before we [released version 2.0](https://github.com/getsentry/sentry-python/releases/tag/2.0.0).
+
+On the first day, there was a [bug report](https://github.com/getsentry/sentry-python/issues/3021) from a user when using the SDK with the Starlette framework and Uvicorn as the server. We fixed the problem and released [2.0.1](https://github.com/getsentry/sentry-python/releases/tag/2.0.1) right away.
+
+Since then the bug tracker has been quiet, and no other regressions were reported.
```
Comment (sentrivana) [inline data/blog/how-to-refactor-and-not-break-things.md:95]: Is this still the case? Not sure if the newest Celery issue was a 2.0 regression

File: data/blog/how-to-refactor-and-not-break-things.md [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+
+We updated our internal usage dashboards so we could see the adoption of the new release candidate.
+
+## Phase IV: Load Testing and Dogfooding
+
+We started to do load testing. We created a [sample application and ran load tests](https://github.com/getsentry/demo-flask-load-test) locally to check if the CPU and memory usage would change between the current SDK and the new 2.0 version. It did not. After the load tests, we were confident that we could use the release candidate on Sentry.io.
+
+Sentry.io is a very big Python project where we use all the advanced features of the SDK. It is a perfect candidate for dogfooding because it will uncover problems very soon.
+
+We first installed the new SDK on our Canary servers. We let it run for an hour and closely monitored CPU and Memory. Everything looked good. Like with the local load tests, nothing spiked.
```
Comment (sentrivana) [inline data/blog/how-to-refactor-and-not-break-things.md:81]:
```suggestion
We first installed the new SDK on our Canary servers. We let it run for an hour and closely monitored CPU and memory. Everything looked good. Like with the local load tests, nothing spiked.
```

**Review by sentrivana — APPROVED** [review]
Looking great! Left a couple of comments

File: data/blog/how-to-refactor-and-not-break-things.md (hunk 11 above)
Comment (antonpirker) [reply data/blog/how-to-refactor-and-not-break-things.md:97]: yea, better not mention the name.

File: data/blog/how-to-refactor-and-not-break-things.md (hunk 5 above)
Comment (antonpirker) [reply data/blog/how-to-refactor-and-not-break-things.md:59]: true. good point. will rephrase.

File: data/blog/how-to-refactor-and-not-break-things.md (hunk 12 above)
Comment (antonpirker) [reply data/blog/how-to-refactor-and-not-break-things.md]: rephrased it and updated the numbers

File: data/blog/how-to-refactor-and-not-break-things.md (hunk 6 above)
Comment (antonpirker) [reply data/blog/how-to-refactor-and-not-break-things.md]: Good idea! Added a code snippet and links to prs!

**Review by bruno-garcia — APPROVED** [review]
